Data-Structures/Linked-List/CycleDetectionII.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * A LinkedList based solution for Detecting a start node of the Cycle in a list.
+ * https://en.wikipedia.org/wiki/Cycle_detection
+ */
+
+function detectCycleNode(head) {
+  /*
+  Problem Statement:
+  Given the head of a linked list, return the node where the cycle begins. If there is no cycle, return null.
+  Link for the Problem: https://leetcode.com/problems/linked-list-cycle-ii/
+  */
+
+  let length = 0
+  let fast = head
+  let slow = head
+
+  while (fast !== null && fast.next !== null) {
+    fast = fast.next.next
+    slow = slow.next
+    if (fast === slow) {
+      // Calculate length of the cycle
+      length = lengthCycle(slow)
+      break
+    }
+  }
+
+  if (length === 0) {
+    // If length = 0, return null
+    return null
+  }
+
+  //Point the nodes at head
+  fast = head
+  slow = head
+  // Move slow pointer ahead by 'length' of cycle times
+  while (length > 0) {
+    slow = slow.next
+    length--
+  }
+
+  // Now move both pointers until they meet - this will be the start of cycle
+  while (fast !== slow) {
+    fast = fast.next
+    slow = slow.next
+  }
+
+  // return the meeting node
+  return slow
+}
+
+function lengthCycle(head) {
```

Comment:
This function seems overcomplicated to me. This function is only called if there is a cycle, yet there is effectively dead/unnecessary code to handle the case that there is no cycle. Also, why do you need another "tortoise-and-hare" here? If `head` already is a node on the cycle, can't you just determine the length of the cycle by following pointers until you're back to `head`? (Effectively your inner loop.)

If I'm missing something and this additional work is necessary, there should be comments justifying / explaining its purpose. It is not self-evident.

Reply:
Thank you for pointing out the potential overcomplication. Though it appears redundant, the inner loop and both pointers are crucial for calculating the cycle length accurately. Even if the head is already on the cycle, we need both the pointers to actually go inside the cycle to determine its length. The condition you mentioned isn't meant to be explicitly checked. It automatically falls into place as both pointers move within the cycle. Also to make the 'length' function more universal or like isolated, I've kept it independent of other functions and reusable.

Reply:
> Though it appears redundant, the inner loop and both pointers are crucial for calculating the cycle length accurately.

I never challenged the necessity of the inner loop, quite the opposite.

> Even if the head is already on the cycle, we need both the pointers to actually go inside the cycle to determine its length.

How so? As said, just the inner loop should work.

> Also to make the 'length' function more universal or like isolated, I've kept it independent of other functions and reusable.

This is pointless given that it's not exported.

Even if it was exported (and tested), I don't think duplicating code and responsibilities is a good idea. Finding a node on a cycle and finding the length of the cycle are two different things. There is no need to duplicate the "finding a node on a cycle" (tortoise and hare) logic.

To make sure that my logic is correct, I simplified the `lengthCycle` function to

```javascript
// head is a node on a cycle
function lengthCycle(head) {
  // How long until we visit head again?
  let cur = head;
  let len = 0;
  do {
  	cur = cur.next;
  	len++;
  } while (cur != head);
  return len;
}
```

and as expected, the tests are still passing.